### DISCUSSION TITLE
Cash on delivery - Allow creating orders with the payment status of `NOT_CHARGED`

### DISCUSSION THREAD

**Discussion by Sagital** [saleor/saleor#14429] [Ideas]
I know this issue has been discussed 2 years ago  https://github.com/mirumee/saleor/issues/3564 but that issue is closed and I am wondering if there is new information about it. 

### What I'm trying to achieve

Currently the checkout flow only allows orders to be created if they are fully paid.  However there are at least two situations when the client won't pay for the order at creation time.

  * Cash On Delivery - a common scenario where the user pays when he receives the goods
  * Bank Transfer - when the user will pay independently of the checkout process, also common in business to business scenarios ( the client pays after a fixed period of time).

### Describe a proposed solution
Orders without payments should be able to be created with a `paymentStatus` of `NOT_CHAGED` . This might be a configuration flag in order to maintain backwards compatibility.

### Other solutions I've tried and won't work

I can use a dummy gateway and fake charge the whole sum before creating the order but this creates some complications down the road, and it also sends an email to the client about the payment which would be kinda confusing (don't know if this is configurable though).

**Comment by dsychin** [comment]
I am also trying to implement something similar and can't figure out a good solution for this.

**Comment by gorjan-mishevski** [comment]
This would be an awesome feature to have definitely. Especially in developing countries where online purchases are still not trusted.

**Comment by krzysztofwolski** [comment]
Such a feature will for sure land in Saleor, but not 3.0. I would love to see this in 3.1, but now it's too early to know how priorities would look like.

There are already started feature refining sessions for Click and collect (client choose to pick up order himself in the store), so payment on pickup should be the next step.

If there's anyone looking for Cash on delivery - we would appreciate more info if there are existing payment gateways that support such features and if there are any other requirements, like additional fields or data that needs to be gathered from the client before creating an order.

**Comment by krzysztofwolski** [comment]
Regarding Banking Transfer - Adyen gateway supports that: https://www.adyen.com/payment-methods#pmx=pay-by-links

**Comment by Sagital** [comment]
Regarding Cash on Delivery, usually the carries offer this service so it will differ from carrier to carrier.  At a minimum it should have an order reference like an order id.

As for the Banking Transfer, I meant the situations when the store will accept the order and deliver only after it has payment confirmation OR deliver it right away because he has a business relationship with that client which pays them after a specified period. 

I think it would be preferably not to enforce a specific gateway because both the client and seller can have a lot of banks between them.

**Comment by Reef3rm4n** [comment]
I've implemented a plugin for cash on delivery if u want directions

**Reply by anjanikant1999** [reply]
hey @Reef3rm4n  can you please share the disrections for implimenting COD in saleor

**Comment by Reef3rm4n** [comment]
> 
> 
> Regarding Banking Transfer - Adyen gateway supports that: https://www.adyen.com/payment-methods#pmx=pay-by-links

Adyen only accepts big business with many thousands $$ in transactions x month.

**Comment by gorjan-mishevski** [comment]
@JayBenato directions would be very appreciated, haven't started working on this yet, was it hard to implement?

**Comment by Reef3rm4n** [comment]
took me a few hours, some code on the backend and some copy and paste in the front end

**Comment by Reef3rm4n** [comment]
> 
> 
> @JayBenato directions would be very appreciated, haven't started working on this yet, was it hard to implement?

start with understanding the checkout flow... its well described in the documentation and also have a look at some of the payment gateways. Look at the Dummy payment method you wanna do something like that.

**Comment by gorjan-mishevski** [comment]
thanks, it seems like a solution without nasty edge cases, which is good news.

**Comment by Reef3rm4n** [comment]
i will submit a mr for this in the meanwhile if you want i can share the code, you can contact me pvt

**Comment by gorjan-mishevski** [comment]
thanks for the offer, I am still not there yet, but the pr would be great to see if you open one as an example to build upon, not only for me but I think for lots of people that stumble upon here.

**Comment by onecrayon** [comment]
We are also working to implement COD and support for other non-credit card transactions. Our particular use-case is that an external service (or external software) handles the actual payment, and what we care about is if the *invoice* has been generated. So for instance, a B2B customer might make a purchase through our Saleor instance, choosing COD. This will cause a call into our external service. Someone in the office will then process the transaction, generate an invoice, and pick the inventory for shipment. Upon generation of the invoice, the external service will call into Saleor, which will need to mark the order as paid without sending emails to that effect (since Saleor will not know, and should not care, if funds have actually been collected; that's handled externally).

To be clear, we're comfortable implementing support for this by hand using custom plug-ins and such, but I wanted to share the usecase in hopes that it will be helpful designing generic COD support in Saleor core.

**Comment by joshsgoldstein** [comment]
> took me a few hours, some code on the backend and some copy and paste in the front end

@JayBenato I am also interested in this use case as well.. Is there a pr open or we can figure out a way to connect and you can share how you did it

**Comment by Reef3rm4n** [comment]
@joshsgoldstein there should be a dummy payment method, which is a way to mock the payment ... you can do the same and just set the correct state for the payment it self

**Comment by sfehlandt** [comment]
You can do this:

1. Create your own payment gateway plugin

2. Make the authorize function return a successful `GatewayResponse` with `kind=AUTH` but `action_required=false`. E.g:
```
@require_active_plugin
    def authorize_payment(
        self, payment_information: "PaymentData", previous_value
    ) -> "GatewayResponse":
        """Perform payment authorization."""
        customer_name = "{} {}".format(
            payment_information.billing.first_name,
            payment_information.billing.last_name,
        )
        response = GatewayResponse(
                is_success=True,
                action_required=False,
                kind=TransactionKind.AUTH,
                amount=payment_information.amount,
                currency=payment_information.currency,
                transaction_id=payment_information.token or "",
                error=None,
                payment_method_info=PaymentMethodInfo(
                    name=customer_name,
                    type=self.PLUGIN_ID,
                ),
        )
        return response
```
3. Make the process_payment function call the authorize payment:

```
@require_active_plugin
    def process_payment(
        self, payment_information: "PaymentData", previous_value
    ) -> "GatewayResponse":
        return self.authorize_payment(payment_information, previous_value)
```

4. Make the capture_payment return a dummy successful GatewayResponse:

```
@require_active_plugin
    def capture_payment(
        self, payment_information: "PaymentData", previous_value
    ) -> "GatewayResponse":
        """Perform capture transaction."""
        customer_name = "{} {}".format(
            payment_information.billing.first_name,
            payment_information.billing.last_name,
        )
        response = GatewayResponse(
            is_success=True,
            action_required=False,
            kind=TransactionKind.CAPTURE,
            amount=payment_information.amount,
            currency=payment_information.currency,
            transaction_id=payment_information.token or "",
            error=None,
            payment_method_info=PaymentMethodInfo(
                name=customer_name,
                type=self.PLUGIN_ID,
            ),
        )
        return response
```

With this, when the `checkoutComplete` mutation is executed, the checkout will finish and the Order will have a `paymentStatus: NOT_CHARGED`. Then, on the dashboard you can click the "CAPTURE" button (once you have the money). Also, if you have another application or automatic bank transfer validator you can make them perform the `capturePayment` mutation, which will mark the order a `FULLY_PAID`

**Comment by nadahmed** [comment]
@sfehlandt you my friend saved me a lot of headache! Thank you!

And yes... It worked flawlessly for my use case!

**Reply by anjanikant1999** [reply]
@nadahmed  can you please help me how to integrate it ?

**Comment by sfehlandt** [comment]
> @sfehlandt you my friend saved me a lot of headache! Thank you!
> 
> And yes... It worked flawlessly for my use case!

Glad it worked. Just a quick update, I suggest you use the orderCapture mutation instead of paymentCapture because this triggers some events the paymentCapture doesn't, such as the order-fully-paid emails sent to customers through the UserEmails Plugin.

Cheers!

**Comment by nadahmed** [comment]
> > @sfehlandt you my friend saved me a lot of headache! Thank you!
> > And yes... It worked flawlessly for my use case!
> 
> Glad it worked. Just a quick update, I suggest you use the orderCapture mutation instead of paymentCapture because this triggers some events the paymentCapture doesn't, such as the order-fully-paid emails sent to customers through the UserEmails Plugin.
> 
> Cheers!

Well, Im using Saleor 3.2 commit `c6a1f`. And it does seem to send a Successful Payment Email on mailhog when the payment is captured from the default saleor dashboard. Along with two other email after order fulfillment. One with details and other that says it's fulfilled. Is there something more I should be aware of? This is my first (serious) e-commerce project and I'm working on it solo. Sorry if my questions seem too obvious.

**Comment by sfehlandt** [comment]
Then I think it's ok, I'm using version 3.0 so maybe it's that 😉

**Comment by mouh2020** [comment]
> I've implemented a plugin for cash on delivery if u want directions
> 

Can you provide me this plugin ?

**Comment by uhwf91p** [comment]
> i will submit a mr for this in the meanwhile if you want i can share the code, you can contact me pvt

Dear JayBenato, could you please share the plugin? It seems that it'll take a lot of time to write such a plugin if you're not experienced in saleor.
Thank you in advance!

**Comment by polow** [comment]
> > i will submit a mr for this in the meanwhile if you want i can share the code, you can contact me pvt
> 
> Dear JayBenato, could you please share the plugin? It seems that it'll take a lot of time to write such a plugin if you're not experienced in saleor. Thank you in advance!

That would be great
